---
title: "04A_Phylogenetic_Tree_Construction"
output: html_document
date: "2025-03-31"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center", 
                      # Send figures generated in this file to this folder below
                      fig.path = "../figures/04A_Phylogenetic_Tree_Construction/")
```

# Goals 

The goal of this script is to make a phylogenetic tree so we can use it for phylogenetic community analyses like phylogenetic Hill Numbers and the UniFrac beta-diversity measures. 

1. Load in pre-processed phyloseq object. 
2. Create ASV fasta file from the phyloseq object.
3. Align the 16S Sequences from our fasta file using the software [MAFFT](https://academic.oup.com/mbe/article/30/4/772/1073398).  
4. Create a maximum likelihood tree using the software [FastTree2](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0009490). 
5. Write out the tree file for use in the next step of the analyses. 

## Inputs 
1. Our previously pre-processed phyloseq object that has chloroplasts, mitochondria, reverse complements, and negative controls removed, which is located in `data/03_PreProcessing/raw_preprocessed_physeq`

## Outputs 
1. The **aligned 16S rRNA gene sequences**, which is the input to the phylogenetic tree. We will save our alignment to `data/04_PhylogeneticTree/MAAFT_aligned_ASV.fasta`
2. The **calculated phylogenetic tree file**, which is the input to the phylogenetic tree. We will save our alignment to`data/04_PhylogeneticTree/ASVs_unrooted.tree`

# Setup 

## Timing of this script

Let's record how long this file took to run on the class server, which we will record at the end of the script. 

```{r rmd-start}
# What time did we start running this script? 
start_time <- Sys.time()
```

## Set Seed 
```{r set-seed}
set.seed(20020718)
```

## Load Packages 
```{r load-packages}
# Install phytools
#remotes::install_github("liamrevell/phytools")

# Install ggtree
#devtools::install_github("GuangchuangYu/ggtree")

# Now load them into the session 
pacman::p_load(phytools, ggtree, tidyverse, install = FALSE)
```

## 1. Load Data 

```{r load-physeq}
# Load physeq 
load("data/03_PreProcessing/raw_preprocessed_physeq.RData")
raw_preprocessed_physeq
```

# 2. Write ASV Fasta 

```{r asv-fasta}
# Pull out ASV sequences and ASV names
asv_seq_df <- 
  raw_preprocessed_physeq@tax_table %>% 
  data.frame() %>% 
  dplyr::select(ASV, ASVseqs)

#View(asv_seq_df)

# Add the ">" to make fasta header
asv_seq_df$ASV <- paste0(">", asv_seq_df$ASV)

#View(asv_seq_df)

# Create fasta object
asv_seq_fasta <- c(rbind(asv_seq_df$ASV, asv_seq_df$ASVseqs))
head(asv_seq_fasta)

# Write to a fasta file 
write(asv_seq_fasta, 
      file = "data/04_PhylogeneticTree/preprocessed_ASVs.fasta")

```

# 3. Run Alignment with MAAFT 

```{r run-mafft, engine = 'bash', engine.opts = '-l'}
# Write bash code to run MAFFT
# First provide the path to MAFFT
#export PATH=/programs/mafft/bin:$PATH

#cd ../data/04_PhylogeneticTree
# Where am I? 
#echo "The working directory is $PWD"

# Set a seed for consistency and reproducibility 
#RANDOM=20020728

# Now, actually run MAFFT
#/programs/mafft/bin/mafft --auto preprocessed_ASVs.fasta > MAFFT_aligned_ASVs.fasta
```

# 4. Calculate the phylogenetic tree with FastTree2

```{r run-fasttree2, engine = 'bash', engine.opts = '-l'}
# Where am I? 
#echo "The working directory is $PWD"

# Load fasttree 
# Provide export path to fasttree 
#export PATH=/programs/FastTree-2.1.11:$PATH

# Run Fasttree to generate phylogenetic tree 
# parameters: 
    # -nt = indicates it's a nucleotide alignment
    # -gtr = generalized time reversible substitution model 
    # -fastest speed up the model, reduce memory usage (recommended for datasets that have >50,000)
    # -log = output a log file 
    # input alignment file 
    # specify the output tree file 
#FastTree -nt -gtr -fastest -log FastTree.log MAFFT_aligned_ASVs.fasta > ASVs_unrooted.tree
```

# Final info for Reproducibility 

## Check Render Time
```{r stop-time}
# Take the time now that we are at the end of the script
end_time <- Sys.time()
end_time 

# Echo the elapsed time
elapsed_time <- round((end_time - start_time), 3)
elapsed_time
```

## Session Information

```{r session-info}
# Ensure reproducibility with package version information
devtools::session_info()
```

