---
title: "06_Ordination"
output: html_document
date: "2025-04-21"
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      fig.path = "../figures/06_Ordination/")
```

# Goals 

1. Load in phyloseq data with rooted tree & colors and functions.  
2. Evaluate sequencing depth and remove samples, if needed.  
3. Normalize the read counts between samples.  
4. Calculate community **dis**similarities. Numbers between 0 and 1. If 0, completely similar versus if they are 1, then they're completely dissimilar.   
    a. **Sorensen**: Shared Species as a binary value: Abundance-unweighted 
    b. **Bray-Curtis**: Shared Abundant species: Abundance-weighted
    c. **(Abundance-)Weighted UNIFRAC**: Consider Abundant Species and where they fall on the tree  
5. Run statistics with functions from the vegan R package: 
    a. PERMANOVA with `adonis2()`.
    b. betadispR with `betadisper()` and `permutest()`. 
6. Visualize the community data with two unconstrained Ordinations:  
    a. **PCoA**: Linear Method. Eigenvalue = how much variation is explained by each axis. Choose to view axis 1, 2, 3, etc. and plot them together.  
    b. **NMDS**: Non-linear. Smush multiple Dimensions into 2 or 3 axes. Need to report Stress value (ideally <0.15).  

## Inputs 

1. We will need the `phytree_preprocessed_physeq.RData`, which includes a rooted tree (ideally within the archaea!) that we created in `analysis/04B_Phylogenetic_Tree_Inspection.Rmd`. 

## Outputs 

*This is an exciting analysis!!! We get to answer another one of our scientific questions!*

1. Calculated beta-diversity dissimilarity measures (*e.g.* Sorensen, Bray-Curtis, abundance-unweighted and -weighted UniFrac) across every sample. 
2. Ordination figures (*i.e.,* PCoA/NMDS) to include in the scientific paper that visualize the data as it relates to the scientific question.
3. Statistical tests (*i.e.,* PERMANOVA & betadisper) conveying the measured and quantified changes and patterns in biodiversity.


# Scientific Question 

<span style="color: red;">INTERPRETATION #1:  What is your scientific question and hypotheses? Do they relate to abundance? Presence/absence? Phylogeny?</span>

How does microbial composition change over the days post admission?

- *Null Hypothesis:* Microbial composition (*e.g.* Sorensen, Bray-Curtis, abundance-unweighted and -weighted UniFrac) does not change over the days post admission. 
- *Alternative #1 Hypothesis:* The relative abundance profiles of bacterial communities in burn wounds significantly change over time after hospital admission.
- *Altnerative #2 Hypothesis*: The presence or absence of bacterial taxa significantly shifts across days post admission in burn wound samples.


hypothesis 1 relates to abundance; hypothesis 2 relates more to presence/absence

# Set up 

## Timing of this script

```{r rmd-start}
# What time did we start running this script? 
start_time <- Sys.time()
```

## Set the seed

```{r set-seed}
set.seed(238428)
```

## Load Packages, colors & functions  
```{r load-packages}

pacman::p_load(tidyverse, devtools, phyloseq, patchwork, vegan, 
               install = FALSE)

# Load Colors 
source("code/colors.R")

# Load functions 
source("code/functions.R")
```

# 1. Load in the data 

```{r load-data}
# load phyloseq object
load("data/04_PhylogeneticTree/phytree_preprocessed_physeqs.RData")
midpoint_rooted_physeq

# Intuition check on seq depth
min(sample_sums(midpoint_rooted_physeq))

metadata_df <- 
  midpoint_rooted_physeq %>%
  sample_data() %>%
  data.frame()
```

# Normalizing the Read Depth 

## 2. Explore the Raw Read Counts 

```{r explore-read-counts, fig.width=6, fig.height=3}
# calculate read counts per sample 
raw_TotalSeqs_df <- 
  midpoint_rooted_physeq %>%
  # Calculate the total number of sequences/reads
  sample_sums() %>%
  data.frame()

# Take a look 
head(raw_TotalSeqs_df)

# Rename the column 
colnames(raw_TotalSeqs_df)[1] <- "TotalSeqs"

# add a new column of num_ASVs (RAW, non-noramlized # ASVs)
raw_TotalSeqsASVs_df <- 
  raw_TotalSeqs_df %>%
  mutate(num_ASVs = colSums(otu_table(midpoint_rooted_physeq) > 1))

#View(raw_TotalSeqsASVs_df)

# Plot histogram of seq depth 
rawSeq_histogram <- 
  raw_TotalSeqsASVs_df %>%
  ggplot(aes(x = TotalSeqs)) + 
  geom_histogram(bins = 50) + 
  scale_x_continuous(limits = c(0, 7500)) + 
  labs(title = "Raw Seq Depth Histogram") + 
  theme_bw()

# Plot Seq Depth versus num_ASVs
rawSeq_vs_numASV_plot <- 
  raw_TotalSeqsASVs_df %>%
  ggplot(aes(x = num_ASVs, y = TotalSeqs)) + 
  geom_point() + 
  scale_y_continuous(limits = c(0, 7500)) +
  geom_smooth(method = "lm") + 
  labs(title = "Seq Depth vs # ASVs") + 
  theme_bw()

# is it significant? 
summary(lm(TotalSeqs ~ num_ASVs, data = raw_TotalSeqsASVs_df))

# Put the two plots together.
rawSeq_histogram + rawSeq_vs_numASV_plot +
  plot_annotation(tag_levels = "A")
```

<span style="color: red;">INTERPRETATION #2:  Evaluating the above plots, what did you scale the read counts to in your dataset? Why did you choose that number? Do you think it is sufficient to perform a beta-diversity analysis and accurately compare your data to one another?</span>

The read counts are scaled to the median value, which is 1257-1351 counts. This is because the read counts above 2000 are more scattered, causing less evenness between samples. I think it is sufficient for beta-diversity analysis  because looking at the scaled plot, the read counts are now pretty uniform. Although there are a few with fewer counts, as seen by the lower left column in the histogram, the range in read counts is not that big.


# 3. Scale Read Counts 

```{r scale-reads, fig.width=6, fig.height=3}
min(sample_sums(midpoint_rooted_physeq))

# Scale the reads 
scaled_physeq <- 
  midpoint_rooted_physeq %>%
  scale_reads(round = "matround")

# Look at it 
scaled_physeq

# Look at it more!
#View(data.frame(otu_table(archaeal_rooted_physeq)))
#View(data.frame(otu_table(scaled_physeq)))

# Confirm seq read depth of scaled_physeq 
scaled_TotalSeqs_df <- 
  scaled_physeq %>%
  sample_sums() %>%
  data.frame()

colnames(scaled_TotalSeqs_df)[1] <- "TotalSeqs"

# add a new column of num_ASVs (RAW, non-noramlized # ASVs)
scaled_TotalSeqsASVs_df <- 
  scaled_TotalSeqs_df %>%
  mutate(num_ASVs = colSums(otu_table(scaled_physeq) > 1))

# Plot it! 
scaledSeq_histogram <- 
  scaled_TotalSeqs_df %>%
  ggplot(aes(x = TotalSeqs)) + 
  geom_histogram(bins = 50) + 
  labs(title = "Scaled Seq Depth Histogram") + 
  scale_x_continuous(limits = c(0, 7500)) + 
  theme_bw()

# Scaling always gives us a range! 
# MInimum sequences
min(scaled_TotalSeqs_df$TotalSeqs) 
# Total Range 
range(scaled_TotalSeqs_df)
# How many is that range? 
range_seqs <- (max(scaled_TotalSeqs_df$TotalSeqs) - min(scaled_TotalSeqs_df$TotalSeqs))
range_seqs
# And in terms of percent? 
range_seqs/max(scaled_TotalSeqs_df$TotalSeqs)
#View(scaled_TotalSeqs_df)

# Set the axis min and max for the next plot 
max_y <- max(scaled_TotalSeqs_df$TotalSeqs) + 10
min_y <- min(scaled_TotalSeqs_df$TotalSeqs) - 10

# Now, draw the plot 
scaledSeq_vs_numASV_plot <- 
  scaled_TotalSeqsASVs_df %>%
  ggplot(aes(x = num_ASVs, y = TotalSeqs)) + 
  geom_point() + 
  scale_y_continuous(limits = c(min_y, max_y)) +
  #geom_smooth(method = "lm") + 
  theme_bw() + 
  labs(title = "Scaled: Seq Depth vs # ASVs")

# Put the two plots together.
scaledSeq_histogram + scaledSeq_vs_numASV_plot +
  plot_annotation(tag_levels = "A")
```


<span style="color: red;">INTERPRETATION #3:  Evaluate the scaled read count plots above. Are the total number of sequences across all of your samples in the same general range? What is the minimum and maximum (i.e., the range) of the sequencing depth? How was the relationship with the number of ASVs modified after scaling the read counts?</span>

Yes they are now in the same general range. Read counts are now between 1257-1351. There used to be a positive correlation between num of ASVs and the seq depth, but now seq depth stays around the same across different numbers of ASVs.

# Beta Diversity: Dissimilarity Metrics in Microbial Ecology
# 4. Calculating dissimilarity
Now, let's make distance objects of each of the dissimiliarity measures above! 

```{r calc-dissimilarity}
# Sorensen Dissimiliarty
scaled_sorensen_dist <- phyloseq::distance(scaled_physeq, method = "bray", binary = TRUE)

# What does it look like? 
class(scaled_sorensen_dist)
str(scaled_sorensen_dist)
#head(as.matrix(scaled_sorensen_dist))

# Bray-Curtis Dissimiliarty
scaled_bray_dist <- phyloseq::distance(scaled_physeq, method = "bray", binary = FALSE)

# Abundance-Unweighted UniFrac
scaled_uUnifrac_dist <- phyloseq::distance(scaled_physeq, method = "unifrac")

# Abundance-Weighted UniFrac
scaled_wUnifrac_dist <- phyloseq::distance(scaled_physeq, method = "wunifrac")
```

<span style="color: red;">INTERPRETATION #4:  Which dissimilarity metric is most related to your scientific question? Why or why not?</span>

Sorensen is the most relevant.I would like to know if more species are present/acquired over the days post admission and sorensen focuses on the presence/absence of species regardless of abundance. Unifrac is less relevant because I do not care about the phylogeny of species.

### Bray-Curtis

```{r PERMANOVA-categorical-bray}
# Bray-Curtis
## ALWAYS check and confirm the order of your terms and how they interact with each other.  
bray_station_adonis_terms1 <- adonis2(scaled_bray_dist ~  `collection.day.post.admission` * `env_local_scale`, data = metadata_df, by = "terms")
bray_station_adonis_terms1

bray_station_adonis_terms2 <- adonis2(scaled_bray_dist ~ `env_local_scale` * `collection.day.post.admission`, data = metadata_df, by = "terms")
bray_station_adonis_terms2


## 2. Run with by = "margin" for marginal p-values
bray_station_adonis_margin <- adonis2(scaled_bray_dist ~ `env_local_scale` * `collection.day.post.admission`, data = metadata_df, by = "margin")
bray_station_adonis_margin
```

### Sorensen

```{r PERMANOVA-categorical-sorensen}
# Sorensen
## 1. Run with by = terms for R² values, sensitive to order of variables! 
## ALWAYS check and confirm the order of your terms and how they interact with each other.  
sorensen_station_adonis_terms1 <- adonis2(scaled_sorensen_dist ~ `collection.day.post.admission` * `env_local_scale`, data = metadata_df, by = "terms")
sorensen_station_adonis_terms1

# Check the order of the terms
sorensen_station_adonis_terms2 <- adonis2(scaled_sorensen_dist ~ `env_local_scale` * `collection.day.post.admission`, data = metadata_df, by = "terms")
sorensen_station_adonis_terms2

## 2. Run with by = "margin" for marginal p-values, which we can compare to the residuals from the first one. 
sorensen_station_adonis_margin <- adonis2(scaled_sorensen_dist ~ `env_local_scale` * `collection.day.post.admission`, data = metadata_df, by = "margin")
sorensen_station_adonis_margin
```

<span style="color: red;">INTERPRETATION #5: Using the above example to guide you, run the appropriate PERMANOVA for your scientific question and interpret the results. What biological conclusion can you draw from your PERMANOVA results?</span>

1. Collection day post admission explains ~7% of the variation in microbial community composition, which is statistically significant based on the p-value of < 0.001. This suggests that temporal changes in the community occur during the course of burn wound healing.

2. The site of collection is the strongest factor, explaining ~12% of the variation, with a p-value of < 0.001. It makes sense that microbiome at wounds vs spared skin vs perianal vs rectal are significantly different.

The interaction between collection day and site of collection explains an additional ~7% of the variation, but this interaction is not statistically significant (p = 0.80), meaning that temporal changes in microbial communities appear to be consistent across different locations — i.e., the effect of time does not depend on location.

Finally, ~74% of the variation remains unexplained by these variables, which is typical in complex microbiome datasets.

# 5b. Betadisper: Testing Variances/Dispersions
## Sorensen

```{r betadisper-sorensen}
# Homogeneity of Disperson test with beta dispr
# Sorensen Betadisper - Station 
dispr_sorensen_station <- betadisper(scaled_sorensen_dist, metadata_df$`collection.day.post.admission`)
# permutest() performs a non-parametric permutation test, which is robust and valid for the kind of data used in beta diversity analysis (e.g., dissimilarity matrices).
permutest(dispr_sorensen_station)

# Sorensen Betadisper - Date  
dispr_sorensen_date <- betadisper(scaled_sorensen_dist, metadata_df$`env_local_scale`)
permutest(dispr_sorensen_date)

```

<span style="color: red;">INTERPRETATION #6: Using the above example to guide you, run the appropriate betadisper() and permutest() for your scientific question and interpret the results. What biological conclusion can you draw from your dispersion analysis?</span>

Collection day post admission does not have a significant betadisper()/permutest() result while site of collection does. This means the temporal groups have similar levels of variability while the variation in differs significantly across different sites of collection.

<span style="color: red;">INTERPRETATION #7: Taking into account both the betadisper()/permutest() and the PERMANOVA (adonis2()) test that you ran, what biological conclusion can you draw regarding your data?</span>

Collection day post admission does not have a significant betadisper()/permutest() result. Therefore, the PERMANOVA results likely represent a true difference by time since admission, supporting the alternative hypothesis.

# 6. Visualize Community Dissimilarity with Ordination

# 6a. PCoA: Principal Coordinates Analysis
```{r pcoa-plots, fig.height=3.5, fig.width=7}
### SORENSEN 
# First, calculate PCoA with Soresen
scaled_soren_pcoa <- 
  ordinate(physeq = scaled_physeq,
         method = "PCoA",
         distance = "bray", binary = TRUE)

# Take a quick look
str(scaled_soren_pcoa)

# Plot it: Sorensen PCoA  
sorensen_pcoa_plot <- 
  plot_ordination(physeq = scaled_physeq,
                ordination = scaled_soren_pcoa,
                color = "collection.day.post.admission",
                shape = "env_local_scale",
                title = "Sorensen PCoA") + 
  scale_color_manual(values = colors) + 
  scale_shape_manual(values = c(15, 16, 17,18)) + 
  geom_point(size = 5, alpha = 0.5, aes(color = `collection.day.post.admission`)) + 
  theme_bw() + 
  theme(legend.position = "right")


### Bray-Curtis 
# Second, calculate PCoA with Bray-Curtis
scaled_bray_pcoa <- 
  ordinate(physeq = scaled_physeq,
         method = "PCoA",
         distance = "bray", binary = FALSE)

# Plot it: Bray-Curtis PCoA 
bray_pcoa_plot <- 
  plot_ordination(physeq = scaled_physeq,
                ordination = scaled_bray_pcoa,
                 color = "collection.day.post.admission",
                shape = "env_local_scale",
                title = "Bray-Curtis PCoA") + 
  scale_color_manual(values = colors) + 
  scale_shape_manual(values = c(15, 16, 17,18)) + 
  geom_point(size = 5, alpha = 0.5, aes(color = `collection.day.post.admission`)) + 
  theme_bw() + 
  theme(legend.position = "right")

# Show the plots 
sorensen_pcoa_plot + bray_pcoa_plot + 
  plot_layout(guides = "collect") +
  plot_annotation(tag_levels = "A")
```
<span style="color: red;">INTERPRETATION #8: Plot your data using a PCoA. Interpret what you can conclude from the PCoA.</span>

I dont see clear clustering based on collection date nor collection site. However, two distinct clusters are observed in the Bray-curtis plot, but it is unclear what the factors driving the difference are.


## 6b. NMDS: Non-Metric Multidimensional Scaling 
```{r soren-nmds}
## SORENSEN 
scaled_soren_nmds <- 
  ordinate(physeq = scaled_physeq,
         method = "NMDS",
         distance = "bray", binary = TRUE)

# Plot it! 
sorensen_nmds_plot <- 
  plot_ordination(physeq = scaled_physeq,
                ordination = scaled_soren_nmds,
                color = "collection.day.post.admission",
                shape = "env_local_scale",
                title = "Sorensen NMDS") + 
  scale_color_manual(values = colors) + 
  scale_shape_manual(values = c(15, 16, 17,18)) + 
  geom_point(size = 5, alpha = 0.5, aes(color = `collection.day.post.admission`)) + 
  theme_bw() + 
  theme(legend.position = "right")

### Bray-Curtis 
# Second, calculate NMDS with Bray-Curtis
scaled_bray_nmds <- 
  ordinate(physeq = scaled_physeq,
         method = "NMDS",
         distance = "bray", binary = FALSE)

# Plot it: Bray-Curtis NMDS 
bray_nmds_plot <- 
  plot_ordination(physeq = scaled_physeq,
                ordination = scaled_bray_nmds,
                color = "collection.day.post.admission",
                shape = "env_local_scale",
                title = "Bray-Curtis NMDS") + 
  scale_color_manual(values = colors) + 
  scale_shape_manual(values = c(15, 16, 17,18)) + 
  geom_point(size = 5, alpha = 0.5, aes(color = `collection.day.post.admission`)) + 
  theme_bw() + 
  theme(legend.position = "right")

# Show the plots 
sorensen_nmds_plot + bray_nmds_plot + 
  plot_layout(guides = "collect") +
  plot_annotation(tag_levels = "A")
```

<span style="color: red;">INTERPRETATION #9: Plot your data using NMDS. Interpret what you can conclude from the NMDS.</span>

Clustering is overall unclear. There may be slight clustering based on collection date, where the earlier samples are cluster a bit more to the bottom and left of the plots.

## All ordinations together!

```{r ordinations, fig.width=7, fig.height=6}
sorensen_pcoa_plot + bray_pcoa_plot + 
sorensen_nmds_plot + bray_nmds_plot + 
  plot_layout(guides = "collect") +
  plot_annotation(tag_levels = "A")
```

<span style="color: red;">INTERPRETATION #10: Plot both the PCoA and NMDS together. Which of the two visualizations work best for your data? Explain your reasoning.</span>

PCoA works better for my data in general since it gives clearer clustering. NMDS works bettwe for answering my specific questions since the pattern based on collection day is a bit more visible in these plots.

# Now what? 

Based on the results above, which metric would you use to write up your paper: Sorensen or Bray-Curtis? 


```{r save-phyloseq-object}

# Now, save midpoint rooted phyloseq object
save(scaled_physeq, 
     file = "data/06_Ordination/scaled_physeq.RData")
```

# Final info for Reproducibility 

## Check Render Time
```{r stop-time}
# Take the time now that we are at the end of the script
end_time <- Sys.time()
end_time 

# Echo the elapsed time
elapsed_time <- round((end_time - start_time), 3)
elapsed_time
```

## Session Information

```{r session-info}
# Ensure reproducibility with package version information
devtools::session_info()
```
