---
title: "07_Composition"
output: html_document
date: "2025-04-23"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE,
                      fig.path = "../figures/07_Composition/")
```

# Goals

In this file, we will perform compositional analysis of our scaled/noramlized/rarefied microbial dataset! 

1. Load in scaled/normalized phyloseq object. 
2. Calculate the relative abundances of taxonomic groups at various levels: 
    A. Phylum
    B. Genus
    C. ASV 
3. Plot them, and narrow in on specific taxnomic group of interest

## Inputs 

1. We will need the `scaled_physeq.RData`, which includes a rooted tree that we created in `analysis/06_Ordination/scaled_physeq.RData`. 

## Outputs 

1. Beautiful visualizations of microbial taxa and how they vary across parameters related to our study: station (categorical) and salinity (continuous).
2. Run some stats! 

# Compositional Data 

Microbial abundance data—like 16S rRNA gene or metagenomics data—are typically **compositional:** they represent relative abundances constrained to a constant total (*e.g.,* percent or proportions). This introduces spurious correlations and other issues if analyzed with traditional statistics. This is a very important limitation to microbial data! 

<span style="color: red;">INTERPRETATION #1: What are the limitations of interpreting microbial abundance from relative (aka. compositional) rather than absolute counts?</span>

# Setup 

## Load Packages 
```{r load-packages}
# Load in the packages 
pacman::p_load(tidyverse, devtools, DT, phyloseq, patchwork, 
               install = FALSE)

# load colors
source("code/colors.R")
```

## 1. Load in Scaled Phyloseq object 

The following phyloseq object contains microbial community composition data in a standardized format. In this case, we’ve already normalized the reads (scaled to 1,942 per sample), which is essential for comparing relative abundances.


```{r load-data}
load("data/06_Ordination/scaled_physeq.RData")

# Look at the data 
scaled_physeq

# Intuition check - scaled at 1,942
min(sample_sums(scaled_physeq))
range(sample_sums(scaled_physeq))
```

# Taxonomic Analysis! 
## A. Phylum 
```{r calc-phylum-df}
# Create a phylum level dataframe
phylum_df <- 
  scaled_physeq %>%
  # Agglomerate all ASV counts within a phylum
  tax_glom(taxrank = "Phylum") %>%
  # Calculate the relative abundance! 
  transform_sample_counts(function(x) {x/sum(x)}) %>%
  # Create a dataframe from phyloseq object
  psmelt() %>%
  # Filter out Phyla < 1 % 
  #dplyr::filter(Abundance > 0.01) %>%
  # fix the order of date
    mutate(`collection.day.post.admission` = fct_relevel(`collection.day.post.admission`, c("day-0", "day-3", "day-7","day-14","day-21","day-28")),
         # Fix the station order 
         `env_local_scale` = fct_relevel(`env_local_scale`, c("burn", "spared",
                                          "rectal", "perianal")))

## What are the phylum abundances? 
phylum_df %>%
  group_by(Phylum) %>%
  summarize(mean_PercAbund = round(mean(Abundance), digits = 4)) %>%
  arrange(-mean_PercAbund) %>%
  datatable()

phylum_df

# Make a list of phyla the top phyla 
top10_phyla <- 
  phylum_df %>%
  group_by(Phylum) %>%
  summarize(mean_PercAbund = mean(Abundance)) %>%
  arrange(-mean_PercAbund) %>%
  head(n = 10) %>%
  pull(Phylum)

top10_phyla
```

<span style="color: red;">INTERPRETATION #2: Which phyla in your dataset have the highest relative abundances? What are their abundances? We will focus on these phyla for plotting below. :)</span>

The top 10 phyla and their abundances are Bacillota 0.3413, Pseudomonadota 0.2617, Bacteroidota 0.2616, Actinomycetota 0.0935, Campylobacterota 0.0242, Fusobacteriota 0.0058, Verrucomicrobiota 0.0049, Thermodesulfobacteriota 0.0027, Synergistota 0.001, Cyanobacteriota 0.0009.    

## Stacked Bar plots 

Visualization helps detect patterns in composition and abundance that statistical models may later test. Stacked bar plots are often used but can obscure individual sample variation and visualize too much data all at once. 

Therefore, we will also explore faceted and jittered boxplots below the bar plots to see sample-level trends more clearly.

```{r phylum-stacked-bar, fig.width=9, fig.height=3.5}
# Stacked Bar Plot With All phyla 
# Plot Phylum Abundances - make sure to load phylum_colors 
stacked_bar_plot<-
phylum_df %>%
  dplyr::filter(Phylum %in% top10_phyla) %>%
  # Warning: It's important to have one sample per x value, 
  # otherwise, it will take the sum between multiple samples
  #dplyr::filter(depth == 0.0) %>%
  #dplyr::filter(fraction == "Whole") %>%
  ggplot(aes(x = `collection.day.post.admission`, y = Abundance, fill = Phylum)) + 
  facet_grid(.~`env_local_scale`) + 
  geom_bar(stat = "identity", color = "black") + 
  labs(title = "Top 10 Phyla") + 
  scale_fill_manual(values = phylum_colors) + 
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1))


ggsave(filename = "figures/07_Composition/phylum_stacked_bar_plot.png",plot=stacked_bar_plot)

```

## Faceted Bar plot 

To help compare the phylum abundance between sample types, we can facet by phylum to better see how the changes occur across the stations, which is masked in the stacked bar plot. It's a little better than the stacked bar plot, however, we can do even better! 

```{r phylum-facet-bar, fig.width=5, fig.height=12}
facet_bar_plot<-
phylum_df %>%
  dplyr::filter(Phylum %in% top10_phyla) %>%
  # Individual sample comparison of surface waters, whole fraction 
  # This will allow us to look at individual samples! 
  # Note that whenever plotting stacked bar plots, you should always look at Individual samples! 
  #dplyr::filter(depth == 0.0) %>%
  #dplyr::filter(fraction == "Whole") %>%
  ggplot(aes(x = `env_local_scale`, y = Abundance, fill = Phylum)) + 
  facet_grid(Phylum~`collection.day.post.admission`, scale = "free") + 
  # add the stacked bar 
  geom_bar(stat = "identity") + 
  # change the colors to be our selected colors 
  scale_fill_manual(values = phylum_colors) + 
  theme_bw() + 
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1))

facet_bar_plot

ggsave(filename = "figures/07_Composition/phylum_facet_bar_plot.png",plot=facet_bar_plot)

### Or combined together: 
combined_plot<-
phylum_df %>%
  dplyr::filter(Phylum %in% top10_phyla) %>%
  ggplot(aes(x =  `env_local_scale`, y = Abundance, fill = Phylum, color = Phylum)) + 
  facet_grid(Phylum~`collection.day.post.admission`, scale = "free") + 
  # add the stacked bar 
  geom_jitter() +
  geom_boxplot(outlier.shape = NA, alpha = 0.5) + 
  # change the colors to be our selected colors 
  scale_fill_manual(values = phylum_colors) + 
  scale_color_manual(values = phylum_colors) + 
  theme_bw() + 
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1))
combined_plot
ggsave(filename = "figures/07_Composition/phylum_combined_plot.png",plot=combined_plot)
```

<span style="color: red;">INTERPRETATION #3: Make the above dotplot and boxplot for your sample types and your most abundant phyla. What *initial* conclusions can you draw regarding the sampled microbial communities as it relates to your scientific question?</span>

For many phyla, the abundances are too low to see trends. Skin samples appeared to have distinct profiles from the rectal/perianal samples, which is expected. Abundances of actinomycetota,bacillota, and pseudomonadota seem to be generally higher in spared skin samples. For burn skin, decrease in microbial abundance is seen for bacillota while an increase in abundance is seen for pseudomonadato.
### A1. Actinomycetota 

```{r phylum-actinos, fig.width=7, fig.height=3.5}
# Narrow in on a specific group
# Actinomycetota - y: abundance, x: collection site, dot plot + boxplot
actino_phylum_site <- 
  phylum_df %>%
  dplyr::filter(Phylum == "Actinomycetota",`collection.day.post.admission`=="day-0") %>%
  # build the plot 
  ggplot(aes(x =  `env_local_scale`, y = Abundance, fill = `env_local_scale`, color = `env_local_scale`))  + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Actinomycetota Phylum - day0") + 
  scale_color_manual(values = colors) + 
  scale_fill_manual(values = colors) + 
    theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
          legend.position = "right")
# Statistically: Kruskall-Wallis followed by a Tukey's Posthoc test
# These are non-parametric (non-normal) stat tests 

# Actinomycetota - y: abundance, x: collection day, dot plot + boxplot
actino_phylum_day <- 
  phylum_df %>%
  dplyr::filter(Phylum == "Actinomycetota", env_local_scale=="burn") %>%
  # build the plot 
  ggplot(aes(x =  `collection.day.post.admission`, y = Abundance, fill = `collection.day.post.admission`, color = `collection.day.post.admission`))  + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Actinomycetota Phylum - burn only") + 
  scale_color_manual(values = colors) + 
  scale_fill_manual(values = colors) + 
    theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
          legend.position = "right")
# Statistically: Kruskall-Wallis followed by a Tukey's Posthoc test
# These are non-parametric (non-normal) stat tests 


actino_phylum_site + actino_phylum_day + 
  plot_layout(guides = "collect") +
  plot_annotation(tag_levels = "A")
```

### A2. Bacillota 

```{r phylum-bacill, fig.width=7, fig.height=3.5}
# Narrow in on a specific group
# Actinomycetota - y: abundance, x: collection site, dot plot + boxplot
bacillota_phylum_site <- 
  phylum_df %>%
  dplyr::filter(Phylum == "Bacillota",`collection.day.post.admission`=="day-0") %>%
  # build the plot 
  ggplot(aes(x =  `env_local_scale`, y = Abundance, fill = `env_local_scale`, color = `env_local_scale`))  + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Bacillota Phylum - day0") + 
  scale_color_manual(values = colors) + 
  scale_fill_manual(values = colors) + 
    theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
          legend.position = "right")
# Statistically: Kruskall-Wallis followed by a Tukey's Posthoc test
# These are non-parametric (non-normal) stat tests 

# Actinomycetota - y: abundance, x: collection day, dot plot + boxplot
bacillota_phylum_day <- 
  phylum_df %>%
  dplyr::filter(Phylum == "Bacillota", env_local_scale=="burn") %>%
  # build the plot 
  ggplot(aes(x =  `collection.day.post.admission`, y = Abundance, fill = `collection.day.post.admission`, color = `collection.day.post.admission`))  + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Bacillota Phylum - burn only") + 
  scale_color_manual(values = colors) + 
  scale_fill_manual(values = colors) + 
    theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
          legend.position = "right")
# Statistically: Kruskall-Wallis followed by a Tukey's Posthoc test
# These are non-parametric (non-normal) stat tests 


bacillota_phylum_site + bacillota_phylum_day + 
  plot_layout(guides = "collect") +
  plot_annotation(tag_levels = "A")
```

### A3. Bacteroidota

```{r phylum-bactero, fig.width=7, fig.height=3.5}
# Narrow in on a specific group
# Actinomycetota - y: abundance, x: collection site, dot plot + boxplot
Bacteroidota_phylum_site <- 
  phylum_df %>%
  dplyr::filter(Phylum == "Bacteroidota",`collection.day.post.admission`=="day-0") %>%
  # build the plot 
  ggplot(aes(x =  `env_local_scale`, y = Abundance, fill = `env_local_scale`, color = `env_local_scale`))  + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Bacteroidota Phylum - day0") + 
  scale_color_manual(values = colors) + 
  scale_fill_manual(values = colors) + 
    theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
          legend.position = "right")
# Statistically: Kruskall-Wallis followed by a Tukey's Posthoc test
# These are non-parametric (non-normal) stat tests 

# Actinomycetota - y: abundance, x: collection day, dot plot + boxplot
Bacteroidota_phylum_day <- 
  phylum_df %>%
  dplyr::filter(Phylum == "Bacteroidota", env_local_scale=="burn") %>%
  # build the plot 
  ggplot(aes(x =  `collection.day.post.admission`, y = Abundance, fill = `collection.day.post.admission`, color = `collection.day.post.admission`))  + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Bacteroidota Phylum - burn only") + 
  scale_color_manual(values = colors) + 
  scale_fill_manual(values = colors) + 
    theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
          legend.position = "right")
# Statistically: Kruskall-Wallis followed by a Tukey's Posthoc test
# These are non-parametric (non-normal) stat tests 


Bacteroidota_phylum_site + Bacteroidota_phylum_day + 
  plot_layout(guides = "collect") +
  plot_annotation(tag_levels = "A")
```

### A4. Pseudomonadota

```{r phylum-pseudo, fig.width=7, fig.height=3.5}
# Narrow in on a specific group
# Actinomycetota - y: abundance, x: collection site, dot plot + boxplot
Pseudomonadota_phylum_site <- 
  phylum_df %>%
  dplyr::filter(Phylum == "Pseudomonadota",`collection.day.post.admission`=="day-0") %>%
  # build the plot 
  ggplot(aes(x =  `env_local_scale`, y = Abundance, fill = `env_local_scale`, color = `env_local_scale`))  + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Pseudomonadota Phylum - day0") + 
  scale_color_manual(values = colors) + 
  scale_fill_manual(values = colors) + 
    theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
          legend.position = "right")
# Statistically: Kruskall-Wallis followed by a Tukey's Posthoc test
# These are non-parametric (non-normal) stat tests 

# Actinomycetota - y: abundance, x: collection day, dot plot + boxplot
Pseudomonadota_phylum_day <- 
  phylum_df %>%
  dplyr::filter(Phylum == "Pseudomonadota", env_local_scale=="burn") %>%
  # build the plot 
  ggplot(aes(x =  `collection.day.post.admission`, y = Abundance, fill = `collection.day.post.admission`, color = `collection.day.post.admission`))  + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Pseudomonadota Phylum - burn only") + 
  scale_color_manual(values = colors) + 
  scale_fill_manual(values = colors) + 
    theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
          legend.position = "right")
# Statistically: Kruskall-Wallis followed by a Tukey's Posthoc test
# These are non-parametric (non-normal) stat tests 


Pseudomonadota_phylum_site + Pseudomonadota_phylum_day + 
  plot_layout(guides = "collect") +
  plot_annotation(tag_levels = "A")
```

<span style="color: red;">INTERPRETATION #4: Based on these phylum-level analyses, were there any phyla that you were able to identify to focus on moving forward for deeper taxonomic analyses? Why?</span>

I think Actinomycetota and Bacillota are worth further analyzing. For Actinomysetota, on the day of admission, there is a significant difference in abundance between burn and spared skin samples. For Bacillota, there may be a decreasing trend in abundance over the days past admission for burn wounds, and on the day of admission, abundance was also slightly lower in spared skin samples.
## B. Genus

```{r genus, fig.width=12, fig.height=4}
# Calculate the Family relative abundance 
# Note: The read depth MUST be normalized in some way: scale_reads
genus_df <- 
  scaled_physeq %>%
  # agglomerate at the phylum level 
  tax_glom(taxrank = "Genus") %>% 
  # Transform counts to relative abundance 
  transform_sample_counts(function (x) {x/sum(x)}) %>%
  # Melt to a long format 
  psmelt() %>%
  # Fix the order of date
  mutate(`collection.day.post.admission` = fct_relevel(`collection.day.post.admission`, c("day-0", "day-3", "day-7","day-14","day-21","day-28")),
         # Fix the station order 
         `env_local_scale` = fct_relevel(`env_local_scale`, c("burn", "spared",
                                          "rectal", "perianal")))


```

### B1. Actinomycetota Genera 

```{r actino-genus}
actino_genus_df<-
  genus_df%>%
  dplyr::filter(Phylum =="Actinomycetota")
actino_genus_df %>%
  group_by(Genus) %>%
  summarize(mean_PercAbund = round(mean(Abundance), digits = 4)) %>%
  arrange(-mean_PercAbund) %>%
  datatable()
top10_actino_genera <- 
  actino_genus_df %>%
  group_by(Genus) %>%
  summarize(mean_PercAbund = mean(Abundance)) %>%
  arrange(-mean_PercAbund) %>%
  head(n = 10) %>%
  pull(Genus)

top10_actino_genera

actino_genus_10<-
  actino_genus_df %>%
  dplyr::filter(Genus %in% top10_actino_genera)%>%
  ggplot(aes(x =  `env_local_scale`, y = Abundance, fill = Genus, color = Genus)) + 
  facet_grid(Genus~`collection.day.post.admission`, scale = "free") + 
  # add the stacked bar 
  geom_jitter() +
  geom_boxplot(outlier.shape = NA, alpha = 0.5) + 
  # change the colors to be our selected colors 
  #scale_fill_manual(values = phylum_colors) + 
  #scale_color_manual(values = phylum_colors) + 
  theme_bw() + 
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1))
  
actino_genus_10

```


```{r actino-genus-plot, fig.width=6, fig.height=6}
# Actinobacteriota
# Plot genus 
actino_genus_site <- 
  actino_genus_df %>%
  dplyr::filter(`collection.day.post.admission`=="day-0") %>%
  # build the plot 
  # At first, plot all of the genera and then subset the ones that have intersting trends
  dplyr::filter(Genus %in% c("Cutibacterium","Microbacterium")) %>%
  # build the plot 
  ggplot(aes(x =  `env_local_scale`, y = Abundance, fill = `env_local_scale`, color = `env_local_scale`)) + 
  facet_wrap(.~Genus, scales = "free_y", nrow = 1) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Actinomycetota Genera - day0") + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "none")

actino_genus_site
actino_genus_day <- 
  actino_genus_df %>%
  dplyr::filter(env_local_scale=="burn") %>%
  dplyr::filter(Genus %in% c("Cutibacterium","Microbacterium")) %>%
  # build the plot 
  ggplot(aes(x =  `collection.day.post.admission`, y = Abundance, fill = `collection.day.post.admission`, color = `collection.day.post.admission`))  + 
  facet_wrap(.~Genus, scales = "free_y", nrow = 1) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Actinomycetota Genera - burn only") + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "none")

# Collect the Actinomycetota Plots
actino_genus_site / actino_genus_day + 
  plot_layout(guides = "collect") +
  plot_annotation(tag_levels = "A")
```


### B2. Bacillota Genera 

```{r bacill-geuns}
Bacillota_genus_df<-
  genus_df%>%
  dplyr::filter(Phylum =="Bacillota")
Bacillota_genus_df %>%
  group_by(Genus) %>%
  summarize(mean_PercAbund = round(mean(Abundance), digits = 4)) %>%
  arrange(-mean_PercAbund) %>%
  datatable()
top10_Bacillota_genera <- 
 Bacillota_genus_df %>%
  group_by(Genus) %>%
  summarize(mean_PercAbund = mean(Abundance)) %>%
  arrange(-mean_PercAbund) %>%
  head(n = 10) %>%
  pull(Genus)

top10_Bacillota_genera

Bacillota_genus_10<-
  Bacillota_genus_df %>%
  dplyr::filter(Genus %in% top10_Bacillota_genera)%>%
  ggplot(aes(x =  `env_local_scale`, y = Abundance, fill = Genus, color = Genus)) + 
  facet_grid(Genus~`collection.day.post.admission`, scale = "free") + 
  # add the stacked bar 
  geom_jitter() +
  geom_boxplot(outlier.shape = NA, alpha = 0.5) + 
  # change the colors to be our selected colors 
  #scale_fill_manual(values = phylum_colors) + 
  #scale_color_manual(values = phylum_colors) + 
  theme_bw() + 
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1))
  
Bacillota_genus_10

```


```{r baci-genus-plot, fig.width=6, fig.height=6}

# Plot genus 
Bacillota_genus_site <- 
  Bacillota_genus_df %>%
  dplyr::filter(`collection.day.post.admission`=="day-0") %>%
  # build the plot 
  # At first, plot all of the genera and then subset the ones that have intersting trends
  dplyr::filter(Genus %in% c("Staphylococcus","Streptococcus")) %>%
  # build the plot 
  ggplot(aes(x =  `env_local_scale`, y = Abundance, fill = `env_local_scale`, color = `env_local_scale`)) + 
  facet_wrap(.~Genus, scales = "free_y", nrow = 1) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Bacillota Genera - day0") + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "none")

Bacillota_genus_site

Bacillota_genus_day <- 
  Bacillota_genus_df %>%
  dplyr::filter(env_local_scale=="burn") %>%
  dplyr::filter(Genus %in% c("Staphylococcus","Streptococcus")) %>%
  # build the plot 
  ggplot(aes(x =  `collection.day.post.admission`, y = Abundance, fill = `collection.day.post.admission`, color = `collection.day.post.admission`))  + 
  facet_wrap(.~Genus, scales = "free_y", nrow = 1) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Bacillota Genera - burn only") + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "none")

Bacillota_genus_day

# Collect the Actinomycetota Plots
Bacillota_genus_site / Bacillota_genus_day + 
  plot_layout(guides = "collect") +
  plot_annotation(tag_levels = "A")
```


<span style="color: red;">INTERPRETATION #5: Please make some plots of mid-level taxonomic ranks (e.g. Genus, Family, Order) as it relates to your scientific question. Were there any changes in the mid-level taxonomic trends compared to the phylum-level results? What does that tell you about that taxonomic group?</span>

The difference in abundances between burn and spared skin samples on day-0 is more distinct when looking at Staph and Strep genera specifically. This could mean that for that phylum, they are the main drivers of the difference seen.The trend in abundances seen over days post admission for the Baillota phylum is also apparent for these specific genera.

## C. ASV level

```{r ASV-plots, fig.width=12, fig.height=4}
# Calculate the Family relative abundance 
# Note: The read depth MUST be normalized in some way: scale_reads
ASV_df <- 
  scaled_physeq %>%
  # Prune out ASVs that have fewer than 100 counts! 
  ## LOOK AT HOW MANY ARE REMOVED! We scaled to 1,962 reads! 
  prune_taxa(taxa_sums(.) >= 196, .) %>%
  # agglomerate at the phylum level 
  tax_glom(taxrank = "ASV") %>% 
  # Transform counts to relative abundance 
  transform_sample_counts(function (x) {x/sum(x)}) %>%
  # Melt to a long format 
  psmelt() %>%
  # fix the order of date
  mutate(`collection.day.post.admission` = fct_relevel(`collection.day.post.admission`, c("day-0", "day-3", "day-7","day-14","day-21","day-28")),
         # Fix the station order 
         `env_local_scale` = fct_relevel(`env_local_scale`, c("burn", "spared",
                                          "rectal", "perianal")))
```


### C1. Actinomycetota ASVs 

```{r actino-asvs, fig.width=6, fig.height=8}
# Calculate top couple of ASVs 
# Make a list of phyla the top phyla 
top_actino_ASVs <- 
  ASV_df %>%
  dplyr::filter(Phylum == "Actinomycetota") %>%
  group_by(ASV) %>%
  summarize(mean_Abundance = mean(Abundance)) %>%
  dplyr::filter(mean_Abundance > 0.005) %>%
  pull(ASV)

# Actinobacteriota
# Plot ASVs 
actino_asv_site <- 
  ASV_df %>%
  dplyr::filter(ASV %in% top_actino_ASVs,`collection.day.post.admission`=="day-0") %>%
  # build the plot 
  ggplot(aes(x = `env_local_scale`, y = Abundance, 
             fill = `env_local_scale`, color = `env_local_scale`)) + 
  facet_wrap(Genus~ASV, scales = "free_y", nrow = 2) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Actinomycetota ASVs - day0") + 
  scale_color_manual(values = colors) + 
  scale_fill_manual(values = colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "none")


actino_asv_day <- 
  ASV_df %>%
  dplyr::filter(ASV %in% top_actino_ASVs,env_local_scale=="burn") %>%
  # build the plot 
  ggplot(aes(x =  `collection.day.post.admission`, y = Abundance, fill = `collection.day.post.admission`, color = `collection.day.post.admission`))  + 
  facet_wrap(Genus~ASV, scales = "free_y", nrow = 2) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Actinomycetota ASVs - burn only") + 
  scale_color_manual(values = colors) + 
  scale_fill_manual(values = colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "none")

# Collect the Actinomycetota Plots
actino_asv_site / actino_asv_day + 
  plot_layout(guides = "collect") +
  plot_annotation(tag_levels = "A")
```

### C2. Bacillota

```{r Bacillota-asvs, fig.width=6, fig.height=8}
# Calculate top couple of ASVs 
# Make a list of phyla the top phyla 
top_Bacillota_ASVs <- 
  ASV_df %>%
  dplyr::filter(Phylum == "Bacillota") %>%
  group_by(ASV) %>%
  summarize(mean_Abundance = mean(Abundance)) %>%
  dplyr::filter(mean_Abundance > 0.005) %>%
  pull(ASV)

# Actinobacteriota
# Plot ASVs 
Bacillota_asv_site <- 
  ASV_df %>%
  dplyr::filter(ASV %in% top_Bacillota_ASVs,`collection.day.post.admission`=="day-0") %>%
  # build the plot 
  ggplot(aes(x = `env_local_scale`, y = Abundance, 
             fill = `env_local_scale`, color = `env_local_scale`)) + 
  facet_wrap(Genus~ASV, scales = "free_y", nrow = 2) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Bacillota ASVs - day0") + 
  scale_color_manual(values = colors) + 
  scale_fill_manual(values = colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "none")


Bacillota_asv_day <- 
  ASV_df %>%
  dplyr::filter(ASV %in% top_Bacillota_ASVs,env_local_scale=="burn") %>%
  # build the plot 
  ggplot(aes(x =  `collection.day.post.admission`, y = Abundance, fill = `collection.day.post.admission`, color = `collection.day.post.admission`))  + 
  facet_wrap(Genus~ASV, scales = "free_y", nrow = 2) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Bacillota ASVs - burn only") + 
  scale_color_manual(values = colors) + 
  scale_fill_manual(values = colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "none")

# Collect the Bacillota Plots
Bacillota_asv_site / Bacillota_asv_day + 
  plot_layout(guides = "collect") +
  plot_annotation(tag_levels = "A")
```

### C3. Bacteroidota

```{r Bacteroidota-asvs, fig.width=6, fig.height=8}
# Calculate top couple of ASVs 
# Make a list of phyla the top phyla 
top_Bacteroidota_ASVs <- 
  ASV_df %>%
  dplyr::filter(Phylum == "Bacteroidota") %>%
  group_by(ASV) %>%
  summarize(mean_Abundance = mean(Abundance)) %>%
  dplyr::filter(mean_Abundance > 0.005) %>%
  pull(ASV)

# Bacteroidota
# Plot ASVs 
Bacteroidota_asv_site <- 
  ASV_df %>%
  dplyr::filter(ASV %in% top_Bacteroidota_ASVs,`collection.day.post.admission`=="day-0") %>%
  # build the plot 
  ggplot(aes(x = `env_local_scale`, y = Abundance, 
             fill = `env_local_scale`, color = `env_local_scale`)) + 
  facet_wrap(Genus~ASV, scales = "free_y", nrow = 2) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Bacteroidota ASVs - day0") + 
  scale_color_manual(values = colors) + 
  scale_fill_manual(values = colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "none")


Bacteroidota_asv_day <- 
  ASV_df %>%
  dplyr::filter(ASV %in% top_Bacteroidota_ASVs,env_local_scale=="burn") %>%
  # build the plot 
  ggplot(aes(x =  `collection.day.post.admission`, y = Abundance, fill = `collection.day.post.admission`, color = `collection.day.post.admission`))  + 
  facet_wrap(Genus~ASV, scales = "free_y", nrow = 2) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Bacteroidota ASVs - burn only") + 
  scale_color_manual(values = colors) + 
  scale_fill_manual(values = colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "none")

# Collect the Bacteroidota Plots
Bacteroidota_asv_site / Bacteroidota_asv_day + 
  plot_layout(guides = "collect") +
  plot_annotation(tag_levels = "A")
```

### C4. Pseudomonadota

```{r Pseudomonadota-asvs, fig.width=6, fig.height=8}
# Calculate top couple of ASVs 
# Make a list of phyla the top phyla 
top_Pseudomonadota_ASVs <- 
  ASV_df %>%
  dplyr::filter(Phylum == "Pseudomonadota") %>%
  group_by(ASV) %>%
  summarize(mean_Abundance = mean(Abundance)) %>%
  dplyr::filter(mean_Abundance > 0.005) %>%
  pull(ASV)

# Pseudomonadota
# Plot ASVs 
Pseudomonadota_asv_site <- 
  ASV_df %>%
  dplyr::filter(ASV %in% top_Pseudomonadota_ASVs,`collection.day.post.admission`=="day-0") %>%
  # build the plot 
  ggplot(aes(x = `env_local_scale`, y = Abundance, 
             fill = `env_local_scale`, color = `env_local_scale`)) + 
  facet_wrap(Genus~ASV, scales = "free_y", nrow = 2) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Pseudomonadota ASVs - day0") + 
  scale_color_manual(values = colors) + 
  scale_fill_manual(values = colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "none")


Pseudomonadota_asv_day <- 
  ASV_df %>%
  dplyr::filter(ASV %in% top_Pseudomonadota_ASVs,env_local_scale=="burn") %>%
  # build the plot 
  ggplot(aes(x =  `collection.day.post.admission`, y = Abundance, fill = `collection.day.post.admission`, color = `collection.day.post.admission`))  + 
  facet_wrap(Genus~ASV, scales = "free_y", nrow = 2) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Pseudomonadota ASVs - burn only") + 
  scale_color_manual(values = colors) + 
  scale_fill_manual(values = colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "none")

# Collect the Pseudomonadota Plots
Pseudomonadota_asv_site / Pseudomonadota_asv_day + 
  plot_layout(guides = "collect") +
  plot_annotation(tag_levels = "A")
```

<span style="color: red;">INTERPRETATION #6: Were the most abundant ASVs in your dataset represented in your ASV-level plots? If not, please go back and plot them specifically. (Hint: Remember that we ordered our ASV names by abundance across the entire dataset. Therefore, ASV_0001 will be the most abundant, ASV_0007 will be the 7th most abundant and so on.)</span>

Yes, the ASVs in the ASV-level plots are all from the top 50 ASVs.


<span style="color: red;">INTERPRETATION #7: Please make some plots at the ASV level as it relates to your scientific question. Were there any ASVs that were especially related to trends that you saw? Did a specific phylogenetic group all show the same trends (e.g. like the Actinomycetota ASVs above)? Or did the ASVs have various, likely niche-specific responses (similar to the Cyanobacteriota above)??</span>

For Bacillota, the trend observed in phylum plot is mostly seen in ASVs that belong to Strep/Staph genera. This is in agreement with interpretations made from the genera plots. The other ASVs all have very low abundances.

<span style="color: red;">INTERPRETATION #8: Now, do a quick literature or google search and look back in the paper that you may be replicating (if applicable). Does your ASV-level changes represent something that is currently known? Or is this a novel result? What did you learn about this ASV or a group of ASVs in your dataset?</span>

Staphylococcus and streptococcus are both common causes of skin wound infections. This may explain why they were higher in abundance on day 0 in burn wounds and decreased in abundance over the course of recovery.

# Session Information 
For reproducibility 
```{r session_info}
devtools::session_info()
```